File: server/services/protocol.js
// 4-path protocol conversion: openai <-> anthropic, passthrough.
// ponytail: stubs only — route handlers cover the basics; openai<->anthropic
// for a real deployment needs request shape mapping (messages<->contents,
// system, max_tokens, tools). Expand when first real cross-protocol request
// arrives.

export function openaiReqToAnthropic(body) {
  const out = { model: body.model, messages: [] }
  if (body.messages) {
    let sys = null
    const msgs = []
    for (const m of body.messages) {
      if (m.role === 'system') { sys = (sys ? sys + '\n' : '') + m.content; continue }
      msgs.push({ role: m.role === 'assistant' ? 'assistant' : 'user', content: m.content })
    }
    if (sys) out.system = sys
    out.messages = msgs
  }
  if (body.max_tokens) out.max_tokens = body.max_tokens
  if (body.temperature !== undefined) out.temperature = body.temperature
  if (body.stream) out.stream = body.stream
  return out
}

export function anthropicReqToOpenai(body) {
  const out = { model: body.model, messages: [] }
  if (body.system) out.messages.push({ role: 'system', content: body.system })
  if (body.messages) {
    for (const m of body.messages) {
      out.messages.push({ role: m.role === 'assistant' ? 'assistant' : 'user', content: typeof m.content === 'string' ? m.content : JSON.stringify(m.content) })
    }
  }
  if (body.max_tokens) out.max_tokens = body.max_tokens
  if (body.temperature !== undefined) out.temperature = body.temperature
  if (body.stream) out.stream = body.stream
  return out
}

// ponytail: Anthropic content block → OpenAI tool_calls entry.
// tool_use (custom client tools) → type:'function'
// server_tool_use (web_search, code_execution, etc.) → type:'custom'
function blockToToolCall(block) {
  if (block.type === 'tool_use') {
    return {
      id: block.id,
      type: 'function',
      function: { name: block.name, arguments: JSON.stringify(block.input) },
    }
  }
  if (block.type === 'server_tool_use') {
    return {
      id: block.id,
      type: 'custom',
      custom: { name: block.name, input: JSON.stringify(block.input) },
    }
  }
  return null
}

// ponytail: parse Minimax non-standard tool_call XML into OpenAI tool_calls
function parseMinimaxToolCalls(text) {
  const cleaned = text.replace(/<\]minimax\[>\[/g, '')
  const tcMatch = cleaned.match(/<tool_call>(.*?)<\/tool_call>/s)
  if (!tcMatch) return { cleanText: cleaned.trim() || null, toolCalls: [] }
  const toolCalls = []
  let callId = 0
  const invokeRe = /<invoke\s+name="([^"]+)">(.*?)<\/invoke>/gs
  let m
  while ((m = invokeRe.exec(tcMatch[1])) !== null) {
    const name = m[1]
    const body = m[2]
    const args = {}
    // extract <query>text</query> or <parameter name="k">v</parameter>
    const paramRe = /<(query|parameter)(?:\s+name="([^"]*)")?>(.*?)<\/\1>/gs
    let pm
    while ((pm = paramRe.exec(body)) !== null) {
      const key = pm[1] === 'parameter' ? (pm[2] || 'value') : 'query'
      args[key] = pm[3].trim()
    }
    toolCalls.push({
      id: `minimax_${++callId}`,
      type: 'function',
      function: { name, arguments: JSON.stringify(args) },
    })
  }
  const cleanText = cleaned.replace(/<tool_call>.*?<\/tool_call>/s, '').trim() || null
  // also strip dangling XML tags
  const finalText = cleanText
    ? cleanText.replace(/<(\/)?(?:invoke|query|parameter)[^>]*>/g, '').trim() || null
    : null
  return { cleanText: finalText, toolCalls }
}
const STOP_REASON_MAP = {
  end_turn: 'stop',
  max_tokens: 'length',
  tool_use: 'tool_calls',
  refusal: 'content_filter',
}

export function anthropicRespToOpenAI(data) {
  const blocks = data.content || []
  const rawText = blocks
    .filter(b => b.type === 'text')
    .map(b => b.text)
    .join('\n') || null
  // ponytail: parse Minimax inline tool_call XML → OpenAI tool_calls
  const { cleanText, toolCalls: minimaxToolCalls } = rawText
    ? parseMinimaxToolCalls(rawText)
    : { cleanText: null, toolCalls: [] }
  // ponytail: also extract any standard Anthropic tool_use blocks
  const standardToolCalls = blocks
    .map(blockToToolCall)
    .filter(Boolean)
  const toolCalls = [...standardToolCalls, ...minimaxToolCalls]
  const stopReason = STOP_REASON_MAP[data.stop_reason] || 'stop'
  const finishReason = toolCalls.length > 0 ? 'tool_calls' : stopReason
  return {
    id: data.id || '',
    object: 'chat.completion',
    created: Math.floor(Date.now() / 1000),
    model: data.model || '',
    choices: [{
      index: 0,
      message: {
        role: 'assistant',
        content: cleanText,
        ...(toolCalls.length > 0 ? { tool_calls: toolCalls } : {}),
      },
      finish_reason: finishReason,
    }],
    usage: data.usage ? {
      prompt_tokens: data.usage.input_tokens || 0,
      completion_tokens: data.usage.output_tokens || 0,
      total_tokens: (data.usage.input_tokens || 0) + (data.usage.output_tokens || 0),
    } : undefined,
  }
}

// OpenAI chat.completion → Anthropic Messages API shape.
export function openaiRespToAnthropic(data) {
  const choice = (data.choices || [])[0]
  const msg = choice?.message || {}
  return {
    id: data.id || '',
    type: 'message',
    role: 'assistant',
    model: data.model || '',
    content: [{ type: 'text', text: msg.content || '' }],
    stop_reason: choice?.finish_reason === 'stop' ? 'end_turn' : (choice?.finish_reason || 'end_turn'),
    usage: data.usage ? {
      input_tokens: data.usage.prompt_tokens || 0,
      output_tokens: data.usage.completion_tokens || 0,
    } : undefined,
  }
}

export function extractTokensOpenai(usage) {
  return {
    input_tokens: usage?.prompt_tokens ?? 0,
    output_tokens: usage?.completion_tokens ?? 0,
    cached_tokens: usage?.prompt_tokens_details?.cached_tokens ?? 0,
  }
}

export function extractTokensAnthropic(usage) {
  return {
    input_tokens: usage?.input_tokens ?? 0,
    output_tokens: usage?.output_tokens ?? 0,
    cached_tokens: (usage?.cache_read_input_tokens ?? 0) + (usage?.cache_creation_input_tokens ?? 0),
  }
}

// ponytail: buffer text, detect complete Minimax tool_call XML blocks, parse to tool_calls
function emitMinimaxClean(textDelta, buf, out, msgId, ts) {
  buf.t = (buf.t || '') + (textDelta || '')
  // ponytail: detect complete </tool_call> → parse into tool_calls + clean text
  const tcEnd = buf.t.indexOf('</tool_call>')
  if (tcEnd !== -1) {
    // extract the full tool_call block
    const afterEnd = tcEnd + '</tool_call>'.length
    const fullChunk = buf.t.slice(0, afterEnd)
    const remainder = buf.t.slice(afterEnd)
    const { cleanText, toolCalls } = parseMinimaxToolCalls(fullChunk)
    // emit parsed tool calls as deltas
    if (toolCalls.length > 0) {
      out.push(`data: ${JSON.stringify({ id: msgId.v, object: 'chat.completion.chunk', created: ts, model: '', choices: [{ index: 0, delta: { tool_calls: toolCalls.map((tc, i) => ({ index: i, id: tc.id, type: tc.type, function: { name: tc.function.name, arguments: tc.function.arguments } })) }, finish_reason: null }] })}\n\n`)
    }
    buf.t = cleanText ? (cleanText + '\n' + remainder) : remainder
    return
  }
  // ponytail: no complete block yet — emit safe prefix, keep 100-char tail
  const cleaned = buf.t
    .replace(/<\]minimax\[>\[/g, '')
    .replace(/<(\/)?(?:tool_call|tool_calls|invoke|query|parameter)[^>]*>/g, '')
  const cutoff = Math.max(0, cleaned.length - 100)
  if (cutoff > 0) {
    out.push(`data: ${JSON.stringify({ id: msgId.v, object: 'chat.completion.chunk', created: ts, model: '', choices: [{ index: 0, delta: { content: cleaned.slice(0, cutoff) }, finish_reason: null }] })}\n\n`)
  }
  buf.t = cleaned.slice(cutoff)
}

// ponytail: Anthropic SSE → OpenAI SSE streaming converter.
// Parses complete SSE events (separated by \n\n) and maps to OpenAI format.
// toolBlockByIndex tracks which content_block indices belong to tool calls.
export function anthropicSSEToOpenAI(rawText, buffer, msgId, toolBlockByIndex) {
  buffer.s += rawText
  const out = []
  while (true) {
    const idx = buffer.s.indexOf('\n\n')
    if (idx === -1) break
    const event = buffer.s.slice(0, idx)
    buffer.s = buffer.s.slice(idx + 2)
    const lines = event.split('\n')
    let evType = '', dataStr = ''
    for (const line of lines) {
      if (line.startsWith('event: ')) evType = line.slice(7)
      else if (line.startsWith('data: ')) dataStr = line.slice(6)
    }
    if (!dataStr) continue
    let data
    try { data = JSON.parse(dataStr) } catch { continue }

    const ts = Math.floor(Date.now() / 1000)
    switch (evType) {
      case 'message_start':
        msgId.v = data.message?.id || msgId.v
        out.push(`data: ${JSON.stringify({ id: msgId.v, object: 'chat.completion.chunk', created: ts, model: data.message?.model || '', choices: [{ index: 0, delta: { role: 'assistant', content: '' }, finish_reason: null }] })}\n\n`)
        break
      case 'content_block_start': {
        const block = data.content_block
        if (block && (block.type === 'tool_use' || block.type === 'server_tool_use')) {
          const toolType = block.type === 'server_tool_use' ? 'custom' : 'function'
          const tcIndex = data.index
          toolBlockByIndex[tcIndex] = { id: block.id, name: block.name, type: toolType }
          out.push(`data: ${JSON.stringify({ id: msgId.v, object: 'chat.completion.chunk', created: ts, model: '', choices: [{ index: 0, delta: { tool_calls: [{ index: tcIndex, id: block.id, type: toolType, function: { name: block.name, arguments: '' } }] }, finish_reason: null }] })}\n\n`)
        }
        break
      }
      case 'content_block_delta': {
        const delta = data.delta
        const tcIndex = data.index
        if (delta?.type === 'input_json_delta' && toolBlockByIndex[tcIndex]) {
          out.push(`data: ${JSON.stringify({ id: msgId.v, object: 'chat.completion.chunk', created: ts, model: '', choices: [{ index: 0, delta: { tool_calls: [{ index: tcIndex, function: { arguments: delta.partial_json } }] }, finish_reason: null }] })}\n\n`)
        } else if (delta?.type === 'text_delta') {
          emitMinimaxClean(delta.text || '', buffer, out, msgId, ts)
        }
        break
      }
      case 'content_block_stop':
        break
      case 'message_delta': {
        const reason = STOP_REASON_MAP[data.delta?.stop_reason] || 'stop'
        out.push(`data: ${JSON.stringify({ id: msgId.v, object: 'chat.completion.chunk', created: ts, model: '', choices: [{ index: 0, delta: {}, finish_reason: reason }] })}\n\n`)
        break
      }
      case 'message_stop':
        // ponytail: flush remaining minimax text buffer
        if (buffer.t?.trim()) {
          out.push(`data: ${JSON.stringify({ id: msgId.v, object: 'chat.completion.chunk', created: ts, model: '', choices: [{ index: 0, delta: { content: buffer.t }, finish_reason: null }] })}\n\n`)
          buffer.t = ''
        }
        out.push('data: [DONE]\n\n')
        buffer.doneSent = true
        break
    }
  }
  if (out.length === 0) return ''
  if (out.some(s => s.includes('[DONE]'))) buffer.doneSent = true
  return out.join('')
}

File: server/internal/llm/protocol.go
package llm

import (
	"encoding/json"
	"fmt"
	"regexp"
	"strings"
	"time"
)

type TokenUsage struct {
	InputTokens  int
	OutputTokens int
	CachedTokens int
}

func toInt(v interface{}) int {
	switch n := v.(type) {
	case float64: return int(n)
	case int: return n
	case int64: return int(n)
	}
	return 0
}

func ExtractTokensOpenAI(u interface{}) TokenUsage {
	usage, _ := u.(map[string]interface{})
	if usage == nil { return TokenUsage{} }
	get := func(k string) int { if v, ok := usage[k]; ok { return toInt(v) }; return 0 }
	cached := 0
	if d, ok := usage["prompt_tokens_details"].(map[string]interface{}); ok { cached = toInt(d["cached_tokens"]) }
	return TokenUsage{InputTokens: get("prompt_tokens"), OutputTokens: get("completion_tokens"), CachedTokens: cached}
}

func ExtractTokensAnthropic(u interface{}) TokenUsage {
	usage, _ := u.(map[string]interface{})
	if usage == nil { return TokenUsage{} }
	get := func(k string) int { if v, ok := usage[k]; ok { return toInt(v) }; return 0 }
	return TokenUsage{InputTokens: get("input_tokens"), OutputTokens: get("output_tokens"), CachedTokens: get("cache_read_input_tokens") + get("cache_creation_input_tokens")}
}

func OpenaiReqToAnthropic(body map[string]interface{}) map[string]interface{} {
	out := map[string]interface{}{}
	if m, ok := body["model"]; ok { out["model"] = m }
	if msgs, ok := body["messages"].([]interface{}); ok {
		var content []map[string]interface{}
		var sys string
		for _, item := range msgs { msg,_ := item.(map[string]interface{}); role,_ := msg["role"].(string)
			c := ""; if s,ok := msg["content"].(string); ok { c = s } else if b,err := json.Marshal(msg["content"]); err==nil { c = string(b) }
			if role == "system" { if sys != "" { sys += "\n" }; sys += c; continue }
			r := "user"; if role == "assistant" { r = "assistant" }
			content = append(content, map[string]interface{}{"role":r,"content":c}) }
		if sys != "" { out["system"] = sys }
		out["messages"] = content }
	for _, f := range []string{"max_tokens","temperature","stream"} { if v, ok := body[f]; ok { out[f] = v } }
	return out
}

func AnthropicReqToOpenai(body map[string]interface{}) map[string]interface{} {
	out := map[string]interface{}{}
	if m, ok := body["model"]; ok { out["model"] = m }
	msgs := []map[string]interface{}{}
	if s, ok := body["system"].(string); ok && s != "" { msgs = append(msgs, map[string]interface{}{"role":"system","content":s}) }
	if arr, ok := body["messages"].([]interface{}); ok {
		for _, item := range arr { msg,_ := item.(map[string]interface{}); role,_ := msg["role"].(string)
			r := "user"; if role == "assistant" { r = "assistant" }
			c := ""; if s,ok := msg["content"].(string); ok { c = s } else if b,err := json.Marshal(msg["content"]); err==nil { c = string(b) }
			msgs = append(msgs, map[string]interface{}{"role":r,"content":c}) } }
	out["messages"] = msgs
	for _, f := range []string{"max_tokens","temperature","stream"} { if v, ok := body[f]; ok { out[f] = v } }
	return out
}

var stopReasonMap = map[string]string{"end_turn":"stop","max_tokens":"length","tool_use":"tool_calls","refusal":"content_filter"}

func AnthropicRespToOpenAI(data map[string]interface{}, providerName string) map[string]interface{} {
	blocks,_ := data["content"].([]interface{})
	rawText := ""; for _, b := range blocks { block,_ := b.(map[string]interface{}); if t,ok := block["type"].(string); ok && t=="text" { if txt,ok := block["text"].(string); ok { rawText += txt } } }
	isMiniMax := strings.Contains(strings.ToLower(providerName), "minimax")
	cleanText := rawText; var toolCalls []map[string]interface{}
	if rawText != "" && isMiniMax { cleanText, toolCalls = parseMinimaxToolCalls(rawText) }
	for _, b := range blocks { block,_ := b.(map[string]interface{}); if tc := blockToToolCall(block); tc != nil { toolCalls = append(toolCalls, tc) } }
	stopReason := stopReasonMap[fmt.Sprint(data["stop_reason"])]; if stopReason == "" { stopReason = "stop" }
	finishReason := stopReason; if len(toolCalls) > 0 { finishReason = "tool_calls" }
	result := map[string]interface{}{"id":data["id"],"object":"chat.completion","created":time.Now().Unix(),"model":data["model"],"choices":[]map[string]interface{}{{"index":0,"message":map[string]interface{}{"role":"assistant","content":cleanText},"finish_reason":finishReason}}}
	if len(toolCalls)>0 { result["choices"].([]map[string]interface{})[0]["message"].(map[string]interface{})["tool_calls"] = toolCalls }
	if u,ok := data["usage"].(map[string]interface{}); ok { t := ExtractTokensAnthropic(u); result["usage"] = map[string]int{"prompt_tokens":t.InputTokens,"completion_tokens":t.OutputTokens,"total_tokens":t.InputTokens+t.OutputTokens} }
	return result
}

func OpenaiRespToAnthropic(data map[string]interface{}) map[string]interface{} {
	choices,_ := data["choices"].([]interface{}); msg := map[string]interface{}{}; if len(choices)>0 { if c,ok := choices[0].(map[string]interface{}); ok { if m,ok := c["message"].(map[string]interface{}); ok { msg = m } } }
	return map[string]interface{}{"id":data["id"],"type":"message","role":"assistant","model":data["model"],"content":[]map[string]interface{}{{"type":"text","text":msg["content"]}}}
}

func blockToToolCall(block map[string]interface{}) map[string]interface{} {
	t,_ := block["type"].(string); if t != "tool_use" && t != "server_tool_use" { return nil }
	name,_ := block["name"].(string); input,_ := json.Marshal(block["input"])
	if t == "server_tool_use" { return map[string]interface{}{"id":block["id"],"type":"custom","custom":map[string]string{"name":name,"input":string(input)}} }
	return map[string]interface{}{"id":block["id"],"type":"function","function":map[string]string{"name":name,"arguments":string(input)}}
}

var toolCallRe = regexp.MustCompile(`<tool_call>(.*?)</tool_call>`)
var invokeRe   = regexp.MustCompile(`<invoke\s+name="([^"]+)">(.*?)</invoke>`)
var paramRe    = regexp.MustCompile(`<(query|parameter)(?:\s+name="([^"]*)")?>(.*?)</(?:query|parameter)>`)

func parseMinimaxToolCalls(text string) (string, []map[string]interface{}) {
	cleaned := strings.ReplaceAll(text, "<]minimax[>[", "")
	if m := toolCallRe.FindStringSubmatch(cleaned); m != nil {
		var tcs []map[string]interface{}; callID := 0
		for _, inv := range invokeRe.FindAllStringSubmatch(m[1], -1) {
			name := inv[1]; body := inv[2]; args := map[string]string{}
			for _, pm := range paramRe.FindAllStringSubmatch(body, -1) {
				key := "query"; if pm[1] == "parameter" && pm[2] != "" { key = pm[2] }
				args[key] = strings.TrimSpace(pm[3]) }
			callID++; tcs = append(tcs, map[string]interface{}{"id":fmt.Sprintf("minimax_%d",callID),"type":"function","function":map[string]string{"name":name,"arguments":jsonStr(args)}}) }
		c := toolCallRe.ReplaceAllString(cleaned, ""); c = strings.TrimSpace(c); return c, tcs }
	return cleaned, nil
}

func jsonStr(v interface{}) string { b,_ := json.Marshal(v); return string(b) }
